src/common/transformations/src/transformations/common_optimizations/transpose_sinking_utils.cpp

```diff
@@ -0,0 +1,167 @@
+            continue;
+        {
+            TransposeInputsInfo input_info;
+            input_info.transpose = transpose_node;
+            input_info.transpose_const = constant_node;
+            input_info.input_idx = input_idx;
+            return input_info;
+        }
+    }
+
+    return TransposeInputsInfo();
+}
+
+bool IfNodeHasTransposeInputs(const Output<Node>& output) {
+    TransposeInputsInfo inputs_info = GetFirstTransposeInput(output.get_node_shared_ptr());
+    return !inputs_info.isEmpty();
+}
+
+AxisVector ReverseTransposeOrder(const AxisVector& axis_order) {
+    AxisVector out(axis_order.size());
+    for (size_t i = 0; i < axis_order.size(); i++) {
+        out.at(axis_order[i]) = i;
+    }
+    return out;
+}
+
+void SwapOutputNames(Output<Node> output1, Output<Node> output2) {
+    const auto node2_output_names = output2.get_names();
+    output2.set_names(output1.get_names());
+    output1.set_names(node2_output_names);
+}
+
+void SwapFriendlyNames(NodePtr node1, NodePtr node2) {
+    const std::string node2_name = node2->get_friendly_name();
+    node2->set_friendly_name(node1->get_friendly_name());
+    node1->set_friendly_name(node2_name);
+}
+
+void SwapNames(NodePtr node1, NodePtr node2) {
+    SwapFriendlyNames(node1, node2);
+    SwapOutputNames(node1->output(0), node2->output(0));
+}
+
+namespace sink_forward {
+
+// insert input reversed transposes, remove first input tranpose
+void UpdateInputTransposes(NodePtr main_node, TransposeInputsInfo& transpose_input_info) {
+    const auto transpose_axis_order = transpose_input_info.transpose_const->get_axis_vector_val();
+    const auto reversed_traspose_axis_order = ReverseTransposeOrder(transpose_axis_order);
+    const size_t tranpose_input_index = transpose_input_info.input_idx;
+    const auto transpose_element_type = transpose_input_info.transpose_const->get_element_type();
+
+    for (size_t i = 0; i < main_node->get_input_size(); ++i) {
+        auto input_node = main_node->input_value(i);
+        if (i == tranpose_input_index) {
+            auto transpose_parent = input_node.get_node()->input_value(0);
+            main_node->input(i).replace_source_output(transpose_parent);
+        } else {
+            auto new_transpose_const = std::make_shared<Constant>(transpose_element_type,
+                                                                  Shape{reversed_traspose_axis_order.size()},
+                                                                  reversed_traspose_axis_order);
+            auto new_transpose = std::make_shared<Transpose>(input_node, new_transpose_const);
+
+            main_node->input(i).replace_source_output(new_transpose->output(0));
+
+            copy_runtime_info(input_node.get_node_shared_ptr(), {new_transpose, new_transpose_const});
+        }
+    }
+}
+
+void RemoveZeroInputNode(NodePtr main_node) {
+    auto input_node = main_node->input_value(0);
+    auto parent_node = input_node.get_node()->input_value(0);
```

Comment:
what if input_node is Parameter or Constant?
these nodes don't have `-input_value(0)` 
it will cause an exception or segfault